lib/graphql/schema/middleware_chain.rb

```diff
@@ -8,21 +8,15 @@ class MiddlewareChain
       # @return [Array<#call(*args)>] Steps in this chain, will be called with arguments and `next_middleware`
       attr_reader :steps
 
-      # @return [Array] Arguments passed to steps (followed by `next_middleware`)
-      attr_reader :arguments
-
-      def initialize(steps:, arguments:)
-        # We're gonna destroy this array, so copy it:
-        @steps = steps.dup
-        @arguments = arguments
+      def initialize(steps:)
+        @steps = steps
       end
 
       # Run the next step in the chain, passing in arguments and handle to the next step
-      def call(next_arguments = @arguments)
-        @arguments = next_arguments
-        next_step = steps.shift
-        next_middleware = self
-        next_step.call(*arguments, next_middleware)
+      def invoke(index, arguments)
+        next_step = steps[index]
+        call_next = ->(next_args = arguments) { invoke(index + 1, next_args) }
```

Comment:
Due to the [overhead of function literals](https://bugs.ruby-lang.org/issues/12901), I think we'd be better off with a callable object here, eg: 

```ruby 
module CallNext
  def initialize(chain, next_args, next_offset)
    @chain = chain 
    @next_args = next_args 
    @next_offset = next_offset 
  end 
  def call(next_args = @next_args) 
    @chain.call(@next_offset, next_args) 
  end 
end 

# ... 

call_next = CallNext.new(self, next_args, index + 1)
```

Ever since I learned about that overhead, I've been trying to keep them out of performance-sensitive code! What do you think, would that work here?